## Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

## Combining Data

1. Concatenating
2. Joining
3. Merging

### Concatening

To demonstrate this, let’s define some data that you can use to practice combining via the pd.concat() function. Your workflow will be as follows:
1. Create dictionaries of data (to experiment with)
2. Convert the dictionaries into dataframes
3. Concatenate the dataframes

In [9]:
# Define a dictionary containing January 2020 data
data1 = {'customer_id':['6732', '767', '890', '635'],
        'month':['Jan-20', 'Jan-20', 'Jan-20', 'Jan-20'],
        'purchased_meat':[0, 13, 3, 4],
        'purchased_alcohol':[1, 2, 10, 0],
        'purchased_snacks':[10, 5, 1, 7]}

In [10]:
# Define a dictionary containing February 2020 data
data2 = {'customer_id':['6732', '767', '890', '635'],
    'month':['Feb-20', 'Feb-20', 'Feb-20', 'Feb-20'],
    'purchased_meat':[0, 10, 5, 3],
    'purchased_alcohol':[2, 4, 14, 0],
    'purchased_snacks':[15, 3, 2, 6]}

In [11]:
# Convert the dictionary into dataframe
df1 = pd.DataFrame(data1,index=[0,1,2,3])
df2 = pd.DataFrame(data2,index=[0,1,2,3])

In [12]:
df1

customer_id   month  purchased_meat  purchased_alcohol  purchased_snacks
0        6732  Jan-20               0                  1                10
1         767  Jan-20              13                  2                 5
2         890  Jan-20               3                 10                 1
3         635  Jan-20               4                  0                 7

In [13]:
df2

customer_id   month  purchased_meat  purchased_alcohol  purchased_snacks
0        6732  Feb-20               0                  2                15
1         767  Feb-20              10                  4                 3
2         890  Feb-20               5                 14                 2
3         635  Feb-20               3                  0                 6

In [14]:
# Create a list frames containing the dataframes you want to combine (df1 and df2)
frames = [df1, df2]

In [15]:
# Concatenate the list above with the pandas function pd.concat()
df_concat = pd.concat(frames)

In [16]:
df_concat

customer_id   month  purchased_meat  purchased_alcohol  purchased_snacks
0        6732  Jan-20               0                  1                10
1         767  Jan-20              13                  2                 5
2         890  Jan-20               3                 10                 1
3         635  Jan-20               4                  0                 7
0        6732  Feb-20               0                  2                15
1         767  Feb-20              10                  4                 3
2         890  Feb-20               5                 14                 2
3         635  Feb-20               3                  0                 6

In [17]:
df_concat = pd.concat(frames, axis = 1)

In [18]:
df_concat

customer_id   month  purchased_meat  purchased_alcohol  purchased_snacks  \
0        6732  Jan-20               0                  1                10   
1         767  Jan-20              13                  2                 5   
2         890  Jan-20               3                 10                 1   
3         635  Jan-20               4                  0                 7   

  customer_id   month  purchased_meat  purchased_alcohol  purchased_snacks  
0        6732  Feb-20               0                  2                15  
1         767  Feb-20              10                  4                 3  
2         890  Feb-20               5                 14                 2  
3         635  Feb-20               3                  0                 6

#### To recap, the pd.concat() function:
1. Is suitable for rows or columns of the same length.
2. Will place dataframes on top of each other by default (axis = 0).
3. Requires a list as its main argument (this is why the frames list was created first in the example above).

### Merging

In [19]:
data3 = {'customer_id':['6732', '767', '890', '635'],
    'month':['Jan-20', 'Jan-20', 'Jan-20', 'Jan-20'],
    'days_purchased_on':[0, 10, 4, 1]}

In [21]:
df3 = pd.DataFrame(data3,index=[0,1,2,3])

In [22]:
# Two dataframes of different sizes: df1 and df3
# Combine them using the df.merge() function and set the on argument to pivot around a common column between the two
df_merged = df1.merge(df3, on = 'customer_id')

In [23]:
df_merged

customer_id month_x  purchased_meat  purchased_alcohol  purchased_snacks  \
0        6732  Jan-20               0                  1                10   
1         767  Jan-20              13                  2                 5   
2         890  Jan-20               3                 10                 1   
3         635  Jan-20               4                  0                 7   

  month_y  days_purchased_on  
0  Jan-20                  0  
1  Jan-20                 10  
2  Jan-20                  4  
3  Jan-20                  1

“month_x” and “month_y.” This is a result of the “month” column existing in both dataframes. Because you didn’t specify it as a key, like you did the “customer_id” column, it’s duplicated in the final dataframe.

In [24]:
df_merged = df1.merge(df3, on = ['customer_id', 'month'])

In [25]:
df_merged

customer_id   month  purchased_meat  purchased_alcohol  purchased_snacks  \
0        6732  Jan-20               0                  1                10   
1         767  Jan-20              13                  2                 5   
2         890  Jan-20               3                 10                 1   
3         635  Jan-20               4                  0                 7   

   days_purchased_on  
0                  0  
1                 10  
2                  4  
3                  1

#### Checking whether it's a full match with "indicator = True"

1. The indicator = Trueargument will create a new column called “_merge” within the new, merged dataframe
2. A value of both means the key (or keys) you specified exist in both dataframes
3. While a value of left_only or right_only indicates that the key only exists in either the left or right dataframe
4. If every row within the final dataframe returns a value of both, then you know you have a full match between your dataframes

In [26]:
df_merged = df1.merge(df3, on = ['customer_id', 'month'], indicator = True)

In [27]:
df_merged

customer_id   month  purchased_meat  purchased_alcohol  purchased_snacks  \
0        6732  Jan-20               0                  1                10   
1         767  Jan-20              13                  2                 5   
2         890  Jan-20               3                 10                 1   
3         635  Jan-20               4                  0                 7   

   days_purchased_on _merge  
0                  0   both  
1                 10   both  
2                  4   both  
3                  1   both

In [28]:
df_merged['_merge'].value_counts()

_merge
both          4
left_only     0
right_only    0
Name: count, dtype: int64

As always, remember to make new dataframes when running procedures like this so that you don’t overwrite any data that might be necessary later

In [29]:
# Test merge without overwriting
pd.merge(df1,df3, on = ['customer_id', 'month'], indicator = True)

customer_id   month  purchased_meat  purchased_alcohol  purchased_snacks  \
0        6732  Jan-20               0                  1                10   
1         767  Jan-20              13                  2                 5   
2         890  Jan-20               3                 10                 1   
3         635  Jan-20               4                  0                 7   

   days_purchased_on _merge  
0                  0   both  
1                 10   both  
2                  4   both  
3                  1   both

Here, the merge is simply being tested and the results displayed without actually creating the new dataframe

Because you’re not applying the merge to an existing dataframe (or creating a new dataframe), both dataframes need to be included as arguments within the pd.merge() function. This is different from before, when you included the original dataframe before the dot in the merge() function.

## Merging My Instacart Data

In [30]:
path = r'/Users/nataliawijaya/Documents/Bootcamp/Instacart Basket Analysis/02 Data'

In [31]:
path

'/Users/nataliawijaya/Documents/Bootcamp/Instacart Basket Analysis/02 Data'

In [32]:
# Importing orders_products_prior.csv
df_ords_prior = pd.read_csv(os.path.join(path, 'Original Data', 'orders_products_prior.csv'), index_col = False)

In [34]:
# Importing orders_checked.csv
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_checked.csv'), index_col = False)

In [58]:
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [38]:
df_ords.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             1           1   2398795        1             2   
2             2           2    473747        1             3   
3             3           3   2254736        1             4   
4             4           4    431534        1             5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                   2                  8                     NaN  
1                   3                  7                    15.0  
2                   3                 12                    21.0  
3                   4                  7                    29.0  
4                   4                 15                    28.0

In [52]:
df_ords = df_ords.drop(columns = ['Unnamed: 0.1', 'Unnamed: 0'])

In [54]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [55]:
df_ords_prior.shape

(32434489, 4)

In [56]:
df_ords.shape

(3421083, 6)

In [59]:
# Merging df_ords_prior and df_ords (default = inner join)
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [62]:
df_merged_large

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32434484   2977660   206209            13                   1   
32434485   2977660   206209            13                   1   
32434486   2977660   206209            13                   1   
32434487   2977660   206209            13                   1   
32434488   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32434484                 12                     7.0       14197   
32434485                 12                     7.0       38730   
32434486                 12                     7.0       31477   
32434487                 12                     7.0        6567   
32434488                 12                     7.0       22920   

          add_to_cart_order  reordered _merge  
0                         1          0   both  
1                         2          0   both  
2                         3          0   both  
3                         4          0   both  
4                         5          0   both  
...                     ...        ...    ...  
32434484                  5          1   both  
32434485                  6          0   both  
32434486                  7          0   both  
32434487                  8          0   both  
32434488                  9          0   both  

[32434489 rows x 10 columns]

In [61]:
# Checking using outer join
pd.merge(df_ords,df_ords_prior, on = 'order_id', indicator = True, how = 'outer')

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32640693   2977660   206209            13                   1   
32640694   2977660   206209            13                   1   
32640695   2977660   206209            13                   1   
32640696   2977660   206209            13                   1   
32640697    272231   206209            14                   6   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN       196.0   
1                         8                     NaN     14084.0   
2                         8                     NaN     12427.0   
3                         8                     NaN     26088.0   
4                         8                     NaN     26405.0   
...                     ...                     ...         ...   
32640693                 12                     7.0     38730.0   
32640694                 12                     7.0     31477.0   
32640695                 12                     7.0      6567.0   
32640696                 12                     7.0     22920.0   
32640697                 14                    30.0         NaN   

          add_to_cart_order  reordered     _merge  
0                       1.0        0.0       both  
1                       2.0        0.0       both  
2                       3.0        0.0       both  
3                       4.0        0.0       both  
4                       5.0        0.0       both  
...                     ...        ...        ...  
32640693                6.0        0.0       both  
32640694                7.0        0.0       both  
32640695                8.0        0.0       both  
32640696                9.0        0.0       both  
32640697                NaN        NaN  left_only  

[32640698 rows x 10 columns]

In [67]:
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

In [68]:
# Export data to pickle(.pkl)
df_merged_large.to_pickle(os.path.join(path, 'Prepared Data', 'orders_products_combined.pkl'))

## Exercise|

In [69]:
# Importing the orders_products_combined dataframe from the pickle file I just saved (.pkl)
df_ords_prods_comb = pd.read_pickle(os.path.join(path, 'Prepared Data', 'orders_products_combined.pkl'))

In [70]:
# Importing products_checked.csv
df_prods = pd.read_csv(os.path.join(path, 'Prepared Data', 'products_checked.csv'), index_col = False)

#### Check the shape of the imported dataframe (it should be the same as the one you exported—always check! (32,434,489 rows)

In [73]:
df_ords_prods_comb.shape

(32434489, 10)

In [76]:
df_prods.shape

(49672, 6)

In [74]:
df_ords_prods_comb.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both

In [75]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [78]:
df_prods = df_prods.drop(columns = ['Unnamed: 0'])

In [79]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [80]:
df_prods.shape

(49672, 5)

#### Determine a suitable way to combine the orders_products_combined dataframe with your products data set. 
I can use the product_id as a key value to combine both dataframes

In [81]:
# Before I merge the two dataframes I want to drop the "_merge" column in the df_orders_products_combined
df_ords_prods_comb.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both

In [82]:
# Before I merge the two dataframes I want to drop the "_merge" column in the df_orders_products_combined
df_ords_prods_comb = df_ords_prods_comb.drop(columns = ['_merge'])

In [83]:
df_ords_prods_comb.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  
0                     NaN         196                  1          0  
1                     NaN       14084                  2          0  
2                     NaN       12427                  3          0  
3                     NaN       26088                  4          0  
4                     NaN       26405                  5          0

In [84]:
# Recheck the shape before merging
df_ords_prods_comb.shape

(32434489, 9)

In [85]:
# Recheck the shape before merging
df_prods.shape

(49672, 5)

In [86]:
# Check the merging before creating a new merged dataframe
pd.merge(df_ords_prods_comb,df_prods, on = 'product_id', indicator = True)

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32404854                 15                     1.0       43553   
32404855                 11                     3.0       43553   
32404856                  7                     NaN       42338   
32404857                  5                    15.0       42338   
32404858                 14                     NaN       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices _merge  
0               77              7     9.0   both  
1               77              7     9.0   both  
2               77              7     9.0   both  
3               77              7     9.0   both  
4               77              7     9.0   both  
...            ...            ...     ...    ...  
32404854        64              7     3.7   both  
32404855        64              7     3.7   both  
32404856        50             19     6.9   both  
32404857        50             19     6.9   both  
32404858       130             14    10.3   both  

[32404859 rows x 14 columns]

In [90]:
df_merged = df_ords_prods_comb.merge(df_prods, on = 'product_id', indicator = True)

In [91]:
df_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

#### Confirm the results of the merge using the merge flag.

In [92]:
df_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [94]:
df_merged.shape

(32404859, 14)

#### Export the newly created dataframe as orders_products_merged (.pkl)

In [95]:
# Export df_merged as .pkl
df_merged.to_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged.pkl'))